spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -185,116 +193,79 @@ public int selectPartition(Object key, int partitionCount) {
 
 	}
 
-	private final class ContentTypeConvertingInterceptor extends ChannelInterceptorAdapter {
-
-		private final String contentType;
+	private final class ContentTypeConvertingInterceptor
+			extends ChannelInterceptorAdapter {
 
 		private final MimeType mimeType;
 
 		private final boolean input;
 
-		private final Class<?> klazz;
-
 		private final MessageConverter messageConverter;
 
 		private final boolean provideHint;
 
 		private ContentTypeConvertingInterceptor(String contentType, boolean input) {
-			this.contentType = contentType;
 			this.mimeType = MessageConverterUtils.getMimeType(contentType);
 			this.input = input;
-			if (MessageConverterUtils.X_JAVA_OBJECT.includes(this.mimeType)) {
-				this.klazz = MessageConverterUtils
-						.getJavaTypeForJavaObjectContentType(this.mimeType);
-			}
-			else if (this.mimeType.equals(MessageConverterUtils.X_SPRING_TUPLE)) {
-				this.klazz = Tuple.class;
-			}
-			else if (this.mimeType.getType().equals("text") || this.mimeType.getSubtype().equals(
-					"json") || this.mimeType.getSubtype().equals("xml")) {
-				this.klazz = String.class;
-			}
-			else {
-				this.klazz = byte[].class;
-			}
+
 			this.messageConverter = MessageConverterConfigurer.this.compositeMessageConverterFactory
-					.getMessageConverterForType(this.mimeType);
+					.getMessageConverterForAllRegistered();
 			this.provideHint = this.messageConverter instanceof AbstractMessageConverter;
 		}
 
 		@Override
 		public Message<?> preSend(Message<?> message, MessageChannel channel) {
+			// bypass conversion for ErrorMessges
+			if (message instanceof ErrorMessage) {
+				return message;
+			}
+			// bypass conversion for raw bytes
+			if (message.getPayload() instanceof byte[]) {
+				return MessageConverterConfigurer.this.messageBuilderFactory
+						.withPayload(message.getPayload())
+						.copyHeaders(message.getHeaders())
+						.setHeaderIfAbsent(MessageHeaders.CONTENT_TYPE, this.mimeType)
+						.build();
+			}
+
 			Message<?> sentMessage = null;
-			if (this.klazz.isAssignableFrom(message.getPayload().getClass())) {
-				Object contentTypeFromMessage = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
-				if (contentTypeFromMessage == null) {
-					sentMessage = MessageConverterConfigurer.this.messageBuilderFactory
-							.fromMessage(message)
-							.setHeader(MessageHeaders.CONTENT_TYPE, this.contentType)
-							.build();
-				}
-				else {
-					sentMessage = message;
-				}
+			Object converted;
+
+			if (this.input) {
+				return MessageConverterConfigurer.this.messageBuilderFactory
+						.withPayload(message.getPayload())
```

Comment:
Maybe something has missed here?
I only see that this `if(this.input)` bock is fully similar to the previous for the `if (message.getPayload() instanceof byte[]) {`

Reply:
Hum, good point, I really did not notice as I was stripping the method away as functionality was ported to the converters.

Rules are indeed equal:

On the outbound we check if the payload is bytes so we bypass conversion (performance optimization)

On inbound, well we don't care about inbound at all. Perhaps the right approach should not configure this interceptor for inbound, since conversion on inbound happens at the MethodHandler on both `@StreamListener` and `@ServiceActivator`